mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp

```diff
@@ -1069,6 +1069,52 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
   return {};
 }
 
+namespace {
+/// Fold dim of a memref reshape operation to a load into the reshape's shape
+/// operand.
+struct DimOfMemRefReshape : public OpRewritePattern<DimOp> {
+  using OpRewritePattern<DimOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(DimOp dim,
+                                PatternRewriter &rewriter) const override {
+    auto reshape = dim.getSource().getDefiningOp<ReshapeOp>();
+
+    if (!reshape)
+      return rewriter.notifyMatchFailure(
+          dim, "Dim op is not defined by a reshape op.");
+
+    if (dim.getIndex().getParentBlock() == reshape->getBlock()) {
+      if (auto *definingOp = dim.getIndex().getDefiningOp()) {
+        if (reshape->isBeforeInBlock(definingOp))
+          return rewriter.notifyMatchFailure(
+              dim,
+              "dim.getIndex is not defined before reshape in the same block.");
+      } // else dim.getIndex is a block argument to reshape->getBlock
+    } else if (!dim.getIndex().getParentRegion()->isProperAncestor(
```

Comment:
A shortcut condition before this one can be `reshape->getParentBlock() == dim->getParentBlock()`

Reply:
I am not following how this check will help. Can you elaborate?

Reply:
I believe you can write:
```
} else if (reshape->getParentBlock() =! dim->getParentBlock() && !dim.getIndex().getParentRegion()->isProperAncestor(
```

Because if the reshape and the dim are in the same block, but not the index operation, then we necessarily have the dominance structure you want.